File: 15_dataStructure/04_binaryTree/LinkBinaryTree.c
#include "head.h"

/**
 * 演示基于链表的二叉树的创建，新增，修改，删除，查找，遍历操作
 * 基于链表的二叉树是最主流的二叉树实现方式
 * 
 * 二叉树有以下术语：
 * 
 * 树的层树
 * 树的深度
 * 节点的高度
 * 节点的深度
 */


/**
 * 初始化一个binaryTree。
 * 初始化过程：
 *  采用递归初始化，先初始化根结点，再初始化左子树，初始化左子树的时候一直先初始化左叶子结点，除非输入0或负数才开始
 *  初始化右子树
 */
Node *createNode() {
    int val;
    printf("请输入结点值，要求整数，当输入为0或负数时表示当前结点不创建。\n");
    scanf("%d", &val);

    if (val <= 0) {
        return NULL;
    }
    
    Node *node = (Node*) malloc(sizeof(Node));
    node -> val = val;

    node -> left = createNode();
    node -> right = createNode();

    return node;
}

int main(int argc, char const *argv[]) {

    Node *root = createNode();
    puts("初始化完成！");

    puts("开始前序遍历：");
    preOrder(root);
    puts("");

    puts("开始中序遍历：");
    inOrder(root);
    puts("");

    puts("开始后序遍历：");
    postOrder(root);
    puts("");

    return 0;
}

File: 15_dataStructure/04_binaryTree/java/LinkBinaryTree.java
import java.util.Scanner;

/**
 * 基于链表的二叉树实现
 * 需要实现的操作：
 * 1）初始化二叉树，采用递归方式实现（中序）
 * 2) 遍历（前序 / 中序 / 后序）
 */
public class LinkBinaryTree {

    private Node root;

    public LinkBinaryTree() {}


    /**
     * 初始化二叉树
     * 初始化规则：当输入的值大于0时，表示新建结点，当输入的值小于等于0时，表示结束新建结点
     */
    public Node create() {
        System.out.println("当输入的值大于0时，表示新建结点，当输入的值小于等于0时，表示结束新建结点");

        int val = nextInt();

        Node node;
        if (val <= 0) {
            node = null;
        } else {
            node = new Node();
            node.setVal(val);
            node.setLeft(create());
            node.setRight(create());
        }

        return node;
    }

    /**
     * 前序遍历
     * @param node
     */
    public void preOrder(Node node) {
        if (node != null) {
            System.out.println("node: " + node.getVal());
            preOrder(node.getLeft());
            preOrder(node.getRight());
        }
    }

    /**
     * 中序遍历
     * @param node
     */
    public void inOrder(Node node) {
        if (node != null) {
            inOrder(node.getLeft());
            System.out.println("node: " + node.getVal());
            inOrder(node.getRight());
        }
    }

    /**
     * 后序遍历
     * @param node
     */
    public void postOrder(Node node) {
        if (node != null) {
            postOrder(node.getLeft());
            postOrder(node.getRight());
            System.out.println("node: " + node.getVal());
        }
    }


    private int nextInt() {
        Scanner scanner = new Scanner(System.in);
        // 输入的数值
        return scanner.nextInt();
    }

    public static void main(String[] args) {
        LinkBinaryTree tree = new LinkBinaryTree();
        tree.root = tree.create();

        System.out.println("preOrder: =============");
        tree.preOrder(tree.root);

        System.out.println("inOrder: =============");
        tree.inOrder(tree.root);

        System.out.println("postOrder: =============");
        tree.postOrder(tree.root);
    }
}


/**
 * 二叉树结点
 */

class Node {
    private int val;
    private Node left;
    private Node right;

    public int getVal() {
        return val;
    }

    public void setVal(int val) {
        this.val = val;
    }

    public Node getLeft() {
        return left;
    }

    public void setLeft(Node left) {
        this.left = left;
    }

    public Node getRight() {
        return right;
    }

    public void setRight(Node right) {
        this.right = right;
    }

    @Override
    public String toString() {
        return String.valueOf(val);
    }
}
